src/partition/repart.c

```diff
@@ -3143,6 +3274,136 @@ static int context_mkfs(Context *context) {
+                                .hash_block_size = sector_size,
+                                .salt_size = 32,
+                        });
+        if (r < 0)
+                return log_error_errno(r, "Failed to setup verity hash data: %m");
+
+        r = sym_crypt_get_volume_key_size(cd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine verity root hash size: %m");
+        rhs = (size_t) r;
+
+        rh = malloc(rhs);
+        if (!rh)
+                return log_oom();
+
+        r = sym_crypt_volume_key_get(cd, CRYPT_ANY_SLOT, (char *) rh, &rhs, NULL, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get verity root hash: %m");
+
+        *ret_roothash = TAKE_PTR(rh);
+        *ret_roothash_size = rhs;
+
+        return 0;
+#else
+        return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "libcryptsetup is not supported, cannot setup verity: %m");
+#endif
+}
+
+static int context_verity(Context *context) {
+        int fd = -1, r;
+
+        assert(context);
+
+        LIST_FOREACH(partitions, p, context->partitions) {
+                Partition *dp;
+                _cleanup_(loop_device_unrefp) LoopDevice *hash_device = NULL;
+                LoopDevice *data_device;
+                _cleanup_free_ uint8_t *rh = NULL;
+                size_t rhs = 0; /* Initialize to work around for GCC false positive. */
+
+                if (p->dropped)
+                        continue;
+
+                if (PARTITION_EXISTS(p)) /* Never format existing partitions */
+                        continue;
+
+                if (p->verity != VERITY_HASH)
+                        continue;
+
+                assert_se(dp = find_verity_sibling(context->partitions, p));
+
+                if (dp->dropped)
+                        continue;
+
+                if (PARTITION_EXISTS(dp))
+                        continue;
+
+                if (fd < 0)
+                        assert_se((fd = fdisk_get_devfd(context->fdisk_context)) >= 0);
+
+                r = loop_device_make(fd, O_RDONLY, dp->offset, dp->new_size, 0, LOCK_EX, &data_device);
+                if (r < 0)
+                        return log_error_errno(r,
+                                               "Failed to make loopback device of verity data partition %" PRIu64 ": %m",
+                                               p->partno);
+
+                r = loop_device_make(fd, O_RDWR, p->offset, p->new_size, 0, LOCK_EX, &hash_device);
+                if (r < 0)
+                        return log_error_errno(r,
+                                               "Failed to make loopback device of verity hash partition %" PRIu64 ": %m",
+                                               p->partno);
+
+                r = do_verity_format(data_device, hash_device, context->sector_size, &rh, &rhs);
+                if (r < 0)
+                        return r;
+
+                assert(rhs >= sizeof(sd_id128_t) * 2);
```

Comment:
Yes, but we need it to be at least twice the size to accomodate the boot loader spec, since the first 128 bits need to become the partition uuid for the data partition, and the final 128 bits are needed to become the uuid for the hash partition, that's why I added the assert for twice the size.

Reply:
So, should we also update the condition in dissect_image()??

Reply:
I think so?